packages/roosterjs-editor-dom/lib/selection/experimentalAttributeBasedSelectionMarker.ts

```diff
@@ -0,0 +1,137 @@
+ * create adjacent text nodes, set this parameter to false. This usually happens for undo/redo.
+ * @returns True if selection markers are added, otherwise false.
+ */
+export function markSelection(
+    container: HTMLElement,
+    range: Range,
+    useInlineMarker: boolean
+): boolean {
+    if (!range || queryElements(container, SELECTION_START_SELECTOR).length > 0) {
+        return false;
+    }
+    let start = Position.getStart(range).normalize();
+    let end = Position.getEnd(range).normalize();
+    markPosition(positionToSerializablePosition(start), SELECTION_START_OFFSET_ATTRIBUTE);
+    markPosition(positionToSerializablePosition(end), SELECTION_END_OFFSET_ATTRIBUTE);
+    return true;
+}
+
+/**
+ * If there is selection marker in content, convert into back to a selection range and remove the markers,
+ * otherwise no op.
+ * @param container Container HTML element to query selection markers from
+ * @param retrieveSelectionRange Whether retrieve selection range from the markers if any
+ * @returns The selection range converted from makers, or null if no valid marker found.
+ */
+export function removeMarker(container: HTMLElement, retrieveSelectionRange: boolean): Range {
+    const selectionStart = container.querySelector(SELECTION_START_SELECTOR);
+    const selectionEnd = container.querySelector(SELECTION_END_SELECTOR);
+    if (selectionStart == null || selectionEnd == null) {
+        return null;
+    }
+
+    const toReturn = retrieveSelectionRange
+        ? createRange(
+              serializablePositionToPosition({
+                  anchorElement: selectionStart,
+                  offsetCharacters: Number.parseInt(
+                      selectionStart.getAttribute(SELECTION_START_OFFSET_ATTRIBUTE)
+                  ),
+              }),
+
+              serializablePositionToPosition({
+                  anchorElement: selectionEnd,
+                  offsetCharacters: Number.parseInt(
+                      selectionEnd.getAttribute(SELECTION_END_OFFSET_ATTRIBUTE)
+                  ),
+              })
+          )
+        : null;
+
+    selectionStart.removeAttribute(SELECTION_START_OFFSET_ATTRIBUTE);
+    selectionEnd.removeAttribute(SELECTION_END_OFFSET_ATTRIBUTE);
+
+    return toReturn;
+}
+
+function markPosition(sp: SerializablePosition, offsetAttribute: string) {
+    sp.anchorElement.setAttribute(offsetAttribute, sp.offsetCharacters + '');
+}
+
+function positionToSerializablePosition(p: Position): SerializablePosition {
+    const node = p.node;
+    if (node instanceof Element) {
+        let offsetCharacters = 0;
+        for (let i = 0; i < p.offset; i++) {
+            offsetCharacters += p.element.childNodes[i].textContent.length;
+        }
+        return {
+            anchorElement: node,
+            offsetCharacters,
+        };
+    } else if (node instanceof Text || node instanceof Comment) {
+        let parent = node.parentElement;
```

Comment:
what if you have:

`<div contenteditable="true" id="rooster-div">hello</div>
`
Also, are markers only used for Undo service? Can we return metadata to store along side the snapshot (i.e. non-DOM data).

Reply:
This method doesn't handle root level text elements. For that I think we'd have to come up with an alternate solution.

The only other call I see is in `replaceWithNode`, which doesn't have a clear reason to be using it afaict. @JiuqingSong would know better than I would.

I was thinking about a selection metadata approach as well, but since you are traversing a new DOM tree you have to store the full path required to recover the selection. (e.g. the indexes of each node you need to step down from the root?). That runs into issues if, in `runWithSelectionMarker`, someone is editing the dom.

If we _are_ only using this for Undo, there's no reason we can't store the selection data separately, which seems like a cleaner overall approach to me.